pulp_container/tests/functional/api/test_push_content.py

```diff
@@ -16,313 +17,345 @@
-        namespace = self.namespace_api.list(name="test").results[0]
-        self.addCleanup(self.namespace_api.delete, namespace.pulp_href)
-
-    def test_push_with_view_perms(self):
-        """
-        Test that push is not working if a user has only view permission for push repos.
-        """
-        repo_name = "unsuccessful/perms"
-        local_url = "/".join([self.registry_name, f"{repo_name}:2.0"])
-        image_path = f"{REGISTRY_V2_REPO_PULP}:manifest_a"
-        with self.assertRaises(exceptions.CalledProcessError):
-            self._push(image_path, local_url, self.user_reader)
-
-    def test_push_with_no_perms(self):
-        """
-        Test that user with no permissions can't perform push.
-        """
-        repo_name = "unsuccessful/perms"
-        local_url = "/".join([self.registry_name, f"{repo_name}:2.0"])
-        image_path = f"{REGISTRY_V2_REPO_PULP}:manifest_a"
-        with self.assertRaises(exceptions.CalledProcessError):
-            self._push(image_path, local_url, self.user_helpless)
-
-        # test a user can still pull
-        self._push(image_path, local_url, self.user_creator)
-        with self.assertRaises(exceptions.CalledProcessError):
-            self._push(image_path, local_url, self.user_dist_consumer)
-        self._pull(local_url, self.user_dist_consumer)
-
-        # cleanup, namespace removal also removes related distributions
-        namespace = self.namespace_api.list(name="unsuccessful").results[0]
-        self.addCleanup(self.namespace_api.delete, namespace.pulp_href)
-
-    def test_push_to_existing_namespace(self):
-        """
-        Test the push to existing namespace with collaborator permissions.
-
-        Container distribution perms and manage-namespace one should be enough
-        to push a new distribution.
-        Container distribution perms shouls be enough to push to the existing
-        distribution.
-        """
-        repo_name = "team/owner"
-        local_url = "/".join([self.registry_name, f"{repo_name}:2.0"])
-        image_path = f"{REGISTRY_V2_REPO_PULP}:manifest_a"
-        self._push(image_path, local_url, self.user_creator)
+    assert container_push_repository_api.list(name=repo_name).count == 1
+    with user_creator:
+        assert container_push_repository_api.list(name=repo_name).count == 1
+    with user_dist_collaborator:
+        assert container_push_repository_api.list(name=repo_name).count == 1
+    with user_dist_consumer:
+        assert container_push_repository_api.list(name=repo_name).count == 1
+    with user_namespace_collaborator:
+        assert container_push_repository_api.list(name=repo_name).count == 1
+    with user_reader:
+        assert container_push_repository_api.list(name=repo_name).count == 1
+    # TODO Can someone verify that this repository should be seen by any user?
+    # Is public default?
+    # with user_helpless:
+    #     assert container_push_repository_api.list(name=repo_name).count == 1
```

Comment:
Yes, users with no permissions should see the repository.

I tested this scenario locally. I pushed an image as an administrator. Then, I tried to access the push-repository as a user with no permissions via the Pulp API.

Reply:
I think we have a mess here. It seems like we allow users to see container push repo if the distribution is public, however when it comes to the content queryset, I don't think we show such content

Reply:
we also don't allow "helpless users" to retrieve a specific public-push-repo, just list
```
$ http  https://pulp3-source-fedora34.fluffy.example.com/pulp/api/v3/repositories/container/container-push/ --ignore-netrc --auth pupsik:12345678!
HTTP/1.1 200 OK
Access-Control-Expose-Headers: Correlation-ID
Allow: GET, HEAD, OPTIONS
Connection: keep-alive
Content-Length: 561
Content-Type: application/json
Correlation-ID: df8fac8b819b49c1a68e659abb91e00f
Date: Tue, 12 Apr 2022 10:43:02 GMT
Referrer-Policy: same-origin
Server: nginx/1.20.1
Strict-Transport-Security: max-age=15768000
Vary: Accept, Cookie
X-Content-Type-Options: nosniff
X-Frame-Options: DENY

{
    "count": 1,
    "next": null,
    "previous": null,
    "results": [
        {
            "description": null,
            "latest_version_href": "/pulp/api/v3/repositories/container/container-push/d7adb15a-a7e5-4681-a220-724192275814/versions/5/",
            "manifest_signing_service": null,
            "name": "test/test-repo",
            "pulp_created": "2022-04-12T10:15:37.064498Z",
            "pulp_href": "/pulp/api/v3/repositories/container/container-push/d7adb15a-a7e5-4681-a220-724192275814/",
            "pulp_labels": {},
            "retain_repo_versions": null,
            "versions_href": "/pulp/api/v3/repositories/container/container-push/d7adb15a-a7e5-4681-a220-724192275814/versions/"
        }
    ]
}


(pulp) [vagrant@pulp3-source-fedora34 pulpcore]$ http  https://pulp3-source-fedora34.fluffy.example.com/pulp/api/v3/repositories/container/container-push/d7adb15a-a7e5-4681-a220-724192275814/ --ignore-netrc --auth pupsik:12345678!
HTTP/1.1 403 Forbidden
Access-Control-Expose-Headers: Correlation-ID
Allow: GET, PUT, PATCH, HEAD, OPTIONS
Connection: keep-alive
Content-Length: 63
Content-Type: application/json
Correlation-ID: 3e8e2d524ec84f13b14d8f727b0ab495
Date: Tue, 12 Apr 2022 10:43:08 GMT
Referrer-Policy: same-origin
Server: nginx/1.20.1
Vary: Accept, Cookie
X-Content-Type-Options: nosniff
X-Frame-Options: DENY

{
    "detail": "You do not have permission to perform this action."
}


```

Reply:
As far as i remember, the test still allows helpless to see the repository and i do not understand how.

Reply:
for example here https://github.com/pulp/pulp_container/blob/main/pulp_container/app/viewsets.py#L1147 maybe we should require only `has_namespace_or_obj_perms:container.view_containerdistribution` because this is purely request to Pulp API

Reply:
Good question. I thought we opted for consistency between the two API's. But i agree we should _decide_ eventually.

At least this explains why the test will show that `user_helpless` can see the repo.

Reply:
I agree that we decided to go  with consistency. But I start questioning whether this is worth the additional complexity we are carrying - for example  we have statement - 'I can pull on public repo with podman pull' .In pulp language it is  translated to 'private' flag on the distribution, and right now we are showing only push-repo types that have public distribution but we forgot about 'regular' repos. 
We also forgot to include 'public' distribution when we compose content queryset . By the logic we have, one should be able to pull content from public repo, as well as see it from pulp API. Right now it is not the case https://github.com/pulp/pulp_container/blob/main/pulp_container/app/viewsets.py#L177 
That's one of few oversights and inconsistencies I've noticed, but I am pretty sure we have more :/

Reply:
Btw, I am not suggesting to address these concerns in this PR. But the test will need to be updated once more, when/if we do that.